### ISSUE TITLE
[Bug] Non-deterministic training results despite fixed random seeds

### ISSUE THREAD

**Issue by amadhan882** [keras-team/keras#22620]
## Description

Training the same model twice with identical random seeds, data, and configuration produces significantly different outputs and loss values.

This violates expected determinism when seeds are explicitly fixed.

---

## Minimal Reproduction Example

```python
import keras
from keras import layers
import numpy as np
import tensorflow as tf

# Fix seeds
np.random.seed(42)
tf.random.set_seed(42)

def build_model():
    model = keras.Sequential([
        layers.Dense(16, activation="relu"),
        layers.Dense(4)
    ])
    model.compile(optimizer="adam", loss="mse")
    return model

x = np.random.randn(100, 8).astype("float32")
y = np.random.randn(100, 4).astype("float32")

# Run 1
model1 = build_model()
history1 = model1.fit(x, y, epochs=3, verbose=0)
out1 = model1.predict(x, verbose=0)

# Reset seeds
np.random.seed(42)
tf.random.set_seed(42)

# Run 2
model2 = build_model()
history2 = model2.fit(x, y, epochs=3, verbose=0)
out2 = model2.predict(x, verbose=0)

print("Max output diff:", np.max(np.abs(out1 - out2)))
print("Loss1:", history1.history["loss"])
print("Loss2:", history2.history["loss"])
```

---

## Observed Behavior

```
Max output diff: ~3.97
Loss1 != Loss2
```

---

## Expected Behavior

With fixed seeds and identical inputs, both runs should produce identical (or near-identical) results.

---

## Impact

- Breaks reproducibility guarantees
- Makes experiments non-repeatable
- Affects debugging and evaluation workflows
- Undermines reliability of training results

---

## Notes

The observed difference is significantly larger than expected floating-point noise, indicating a deeper source of non-determinism.

**Comment by amadhan882** [comment]
## Additional verification (CPU-only)

The issue persists even when forcing CPU execution:

```python
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""  # force CPU

import keras
from keras import layers
import numpy as np
import tensorflow as tf

np.random.seed(42)
tf.random.set_seed(42)

def build_model():
    model = keras.Sequential([
        layers.Dense(16, activation="relu"),
        layers.Dense(4)
    ])
    model.compile(optimizer="adam", loss="mse")
    return model

x = np.random.randn(100, 8).astype("float32")
y = np.random.randn(100, 4).astype("float32")

model1 = build_model()
model1.fit(x, y, epochs=3, verbose=0)
out1 = model1(x)

np.random.seed(42)
tf.random.set_seed(42)

model2 = build_model()
model2.fit(x, y, epochs=3, verbose=0)
out2 = model2(x)

print("diff:", np.max(np.abs(out1 - out2)))
```

Result:

```
diff: ~2.75
```

This eliminates GPU nondeterminism as a cause.

The difference remains significantly larger than expected floating-point noise, confirming that the behavior is not due to hardware or parallel execution.

This suggests a deeper source of nondeterminism in the training or initialization pipeline.

**Comment by ChiragSW** [comment]
@amadhan882 please try this ```keras.utils.set_random_seed(42)``` instead. Please let me know if this works, else the problem was using APIs that do not reset Python’s ```random```, which default init depended on.

**Comment by amadhan882** [comment]
@ChiragSW

Thanks for the suggestion — using `keras.utils.set_random_seed(42)` resolves the issue.

It appears the behavior was due to incomplete seeding (Python random was not set).

This can be closed.
